### PR TITLE
cabana: add 0.4 and new dependency variants

### PR DESCRIPTION
**Pull request by streeve**
ping @junghans 

Adds hypre-cmake dependent package (fix #24817)

Fix. https://github.com/ECP-copa/Cabana/issues/427

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 1]
```diff
@@ -11,40 +11,56 @@ class Cabana(CMakePackage):
...
+    version('0.4.0', sha256='c347d23dc4a5204f9cc5906ccf3454f0b0b1612351bbe0d1c58b14cddde81e85')
     version('0.3.0', sha256='fb67ab9aaf254b103ae0eb5cc913ddae3bf3cd0cf6010e9686e577a2981ca84f')
     version('0.2.0', sha256='3e0c0e224e90f4997f6c7e2b92f00ffa18f8bcff72f789e0908cea0828afc2cb')
     version('0.1.0', sha256='3280712facf6932b9d1aff375b24c932abb9f60a8addb0c0a1950afd0cb9b9cf')
     version('0.1.0-rc0', sha256='73754d38aaa0c2a1e012be6959787108fec142294774c23f70292f59c1bdc6c5')
 
-    variant('serial', default=True, description="enable Serial backend (default)")
-    variant('openmp', default=False, description="enable OpenMP backend")
-    variant('cuda', default=False, description="enable Cuda backend")
+    kokkos_backends = {
```
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py]: There is a way to pull `devices_variants` from the kokkos spackage.

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 2]
```diff
@@ -11,40 +11,56 @@ class Cabana(CMakePackage):
...
+        "0.4.0:": "@3.2"
+    }
+    for version in versions:
+        kk_version = versions[version]
+        for backend in kokkos_backends:
+            depends_on('kokkos{0}+{1}'.format(kk_version, backend),
+                       when='@{0}+{1}'.format(version, backend))
+    depends_on("arborx", when="@0.3.0:+arborx")
+    depends_on("heffte@2.0:", when="@0.4.0:+heffte")
+    depends_on("hypre@2.22.0:", when="@0.4.0:+hypre")
```
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py]: We need the cmake variant, right?

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 1 above)
Comment (junghans) [reply var/spack/repos/builtin/packages/cabana/package.py]: Something like:
```
kokkos_pkg = spack.repo.get('kokkos')
kokkos_backends = kokkos_pkg. devices_variants
```

Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Nice! Added

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 2 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Yes, I forgot we removed non-cmake support for hypre. I switched it to `hypre-cmake` which would mean we're blocked by #24817 or other route for cmake support with #25781

**Comment by streeve** [comment]
ECP-copa/Cabana#427

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 3]
```diff
@@ -4,47 +4,58 @@
...
     version('0.3.0', sha256='fb67ab9aaf254b103ae0eb5cc913ddae3bf3cd0cf6010e9686e577a2981ca84f')
     version('0.2.0', sha256='3e0c0e224e90f4997f6c7e2b92f00ffa18f8bcff72f789e0908cea0828afc2cb')
     version('0.1.0', sha256='3280712facf6932b9d1aff375b24c932abb9f60a8addb0c0a1950afd0cb9b9cf')
     version('0.1.0-rc0', sha256='73754d38aaa0c2a1e012be6959787108fec142294774c23f70292f59c1bdc6c5')
 
-    variant('serial', default=True, description="enable Serial backend (default)")
-    variant('openmp', default=False, description="enable OpenMP backend")
-    variant('cuda', default=False, description="enable Cuda backend")
+    kokkos_backends = Kokkos.devices_variants
+    for backend in kokkos_backends:
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]: Use underscores for these class-scope temporary attributes (here and below): `kokkos_backends`, `backend`, `version`, etc.  You definitely need to be careful to not accidentally define `Cabana.version`.

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 4]
```diff
@@ -4,47 +4,58 @@
...
+    depends_on("hypre-cmake@2.22.0:", when="@0.4.0:+hypre")
     depends_on('mpi', when='+mpi')
 
+    conflicts("+rocm", when="@:0.2.0")
+    conflicts("+sycl", when="@:0.3.0")
+
     def cmake_args(self):
         options = [
             '-DCabana_ENABLE_TESTING=ON',
-            '-DCabana_ENABLE_Serial=%s'  % (
```
Comment (junghans) [inline var/spack/repos/builtin/packages/cabana/package.py:40]: Where do these options go?

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 5]
```diff
@@ -4,47 +4,58 @@
...
-                'On' if '+serial'  in self.spec else 'Off'),
-            '-DCabana_ENABLE_OpenMP=%s'  % (
-                'On' if '+openmp'  in self.spec else 'Off'),
-            '-DCabana_ENABLE_Cuda=%s'  % (
-                'On' if '+cuda'  in self.spec else 'Off'),
-            '-DCabana_ENABLE_MPI=%s'  % (
-                'On' if '+mpi'  in self.spec else 'Off'),
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py:47]: Do these removals mean this won't build for older versions? Or were those older versions relying on kokkos?

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 3 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py]: Fixed, thanks

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 4 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:40]: Fixed - I now only define for the previous versions

File: var/spack/repos/builtin/packages/cabana/package.py (hunk 5 above)
Comment (streeve) [reply var/spack/repos/builtin/packages/cabana/package.py:47]: Added back, my mistake. New versions rely on Kokkos, but previous do not

**Review by junghans — DISMISSED** [review]

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 6]
```diff
@@ -59,5 +59,19 @@ def cmake_args(self):
...
+        # These variables were removed in 0.3.0 (where backends are
+        # automatically used from Kokkos)
+        if self.spec.satisfies('@:0.2.0'):
+            backends = {'serial': 'Serial',
+                        'openmp': 'OpenMP',
+                        'cuda': 'Cuda'}
+            for backend in backends:
+                options.append('-DCabana_ENABLE_%s=%s' %
+                               (backends[backend],
+                                'On' if '+' + backend in self.spec else 'Off'))
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]: The `CMakePackage.define_from_variant` will help here (and check to make sure the variant names are valid):
```suggestion
                options.append(self.define_from_variant('Cabana_ENABLE_' + backends[backend], backend))
```

File: var/spack/repos/builtin/packages/cabana/package.py [hunk 7]
```diff
@@ -59,5 +59,19 @@ def cmake_args(self):
...
+
+        if self.spec.satisfies('@:0.3.0'):
+            options.append('-DCabana_ENABLE_MPI=%s' %
+                           ('On' if '+mpi' in self.spec else 'Off'))
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/cabana/package.py]:
```suggestion
            options.append(self.define_from_variant('Cabana_ENABLE_MPI', 'mpi))
```

**Comment by streeve** [comment]
Added `hypre-cmake` from #24817 to pass CI here

**Review by junghans — DISMISSED** [review]

**Review by sethrj — CHANGES_REQUESTED** [review]
@junghans Can hypre-cmake be a separate PR? I think it's nonstandard to create a new package due to a new build system (there are various workarounds in place, and I think the most drastic is labeling the old build system as `-legacy` rather than labeling the new one as `-cmake`).

**Comment by junghans** [comment]
> @junghans Can hypre-cmake be a separate PR? I think it's nonstandard to create a new package due to a new build system (there are various workarounds in place, and I think the most drastic is labeling the old build system as `-legacy` rather than labeling the new one as `-cmake`).

See the discussion in https://github.com/spack/spack/pull/25781#issuecomment-916403334, in short, package rename isn't going to happen soon.

**Comment by junghans** [comment]
Plus we need `hypre-cmake` here to make the CI pass.

**Comment by sethrj** [comment]
> Plus we need `hypre-cmake` here to make the CI pass.

Right, but it's not uncommon to have multiple interdependent prs in Spack.

**Comment by sethrj** [comment]
@junghans if you think it's easier to keep the new hyper-cmake on this branch rather than a new one first, then that's fine and I'll review the new package. (You're both the author and reviewer of that commit so we need another set of eyes on it.)

File: var/spack/repos/builtin/packages/hypre-cmake/package.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    maintainers = ['ulrikeyang', 'osborn9', 'balay']
+
+    test_requires_compiler = True
+
+    version('develop', branch='master')
+    version('2.22.0', sha256='2c786eb5d3e722d8d7b40254f138bef4565b2d4724041e56a8fa073bda5cfbb5')
+
+    variant('shared', default=(sys.platform != 'darwin'),
+            description="Build shared library (disables static library)")
+    variant('superlu-dist', default=False,
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]:
```suggestion
    variant('superlu_dist', default=False,
```
See https://github.com/spack/spack/pull/24631#discussion_r662798100

File: var/spack/repos/builtin/packages/hypre-cmake/package.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+        if '+cuda' in self.spec:
+            args.append('-DHYPRE_WITH_CUDA=ON')
+        else:
+            args.append('-DHYPRE_WITH_CUDA=OFF')
+
+        if '+unified-memory' in self.spec:
+            args.append('-DHYPRE_ENABLE_UNIFIED_MEMORY=ON')
+        else:
+            args.append('-DHYPRE_ENABLE_UNIFIED_MEMORY=OFF')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]:
```suggestion
        from_variant = self.define_from_variant
        args = [
            from_variant('HYPRE_WITH_MPI', 'mpi'),
            from_variant('HYPRE_WITH_OPENMP', 'openmp'),
            from_variant('HYPRE_WITH_BIGINT', 'int64'),
            from_variant('HYPRE_WITH_MIXEDINT', 'mixedint'),
            from_variant('HYPRE_WITH_COMPLEX', 'complex'),
            from_variant('BUILD_SHARED_LIBS', 'shared'),
            from_variant('HYPRE_WITH_DSUPERLU', 'superlu_dist'),
            from_variant('HYPRE_WITH_CUDA', 'cuda'),
            from_variant('HYPRE_ENABLE_UNIFIED_MEMORY', 'unified_memory'),
        ]
```

File: var/spack/repos/builtin/packages/hypre-cmake/package.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    variant('int64', default=False,
+            description="Use 64bit integers")
+    variant('mixedint', default=False,
+            description="Use 64bit integers while reducing memory use")
+    variant('complex', default=False, description='Use complex values')
+    variant('mpi', default=True, description='Enable MPI support')
+    variant('openmp', default=False, description='Enable OpenMP support')
+    variant('debug', default=False,
+            description='Build debug instead of optimized version')
+    variant('unified-memory', default=False, description='Use unified memory')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/hypre-cmake/package.py]: `unified_memory`

**Review by sethrj — CHANGES_REQUESTED** [review]

**Comment by junghans** [comment]
> @junghans if you think it's easier to keep the new hyper-cmake on this branch rather than a new one first, then that's fine and I'll review the new package. (You're both the author and reviewer of that commit so we need another set of eyes on it.)

Thanks for the review.

**Review by sethrj — DISMISSED** [review]
Thanks!

**Review by sethrj — APPROVED** [review]

**Comment by streeve** [comment]
This looks like an unrelated failure - not sure if you can selectively rerun jobs

**Comment by sethrj** [comment]
@spackbot run pipeline

**Pull request merged**